src/main/java/com/bench/stockmanagement/ProductHandler.java

```diff
@@ -1,35 +1,122 @@
+            SellingRepository sellingRepository, ProductRepository productRepository) {
+        this.reader = reader;
         this.productMapper = productMapper;
-        this.orderService = orderService;
-        this.sellingService = sellingService;
+        this.orderRepository = orderRepository;
+        this.sellingRepository = sellingRepository;
+        this.productRepository = productRepository;
+    }
+
+    // Save products
+    public Mono<Result> saveProducts() {
+        List<Order> orders = reader.readOrder();
+        List<DBProduct> dbProducts = productMapper.mapProduct(orders);
+
+        return Flux.fromIterable(dbProducts).log()
+                .doOnNext(productRepository::saveProducts)
+                .then(Mono.just(SUCCESS))
+                .onErrorReturn(FAIL);
+    }
+
+    // Get product by itemNumber
+    // Item number, how many is expected in stock min/max, purchase prices
+    public Mono<List<ProductStockData>> getProductByItemNumber(String itemNumber) {
+        return Mono.fromFuture(productRepository.getProductByItemNumber(itemNumber))
+                .log()
+                .map(productMapper::mapStockData);
+    }
+
+    // Get all information about a product by itemNumber
+    // Name, quantity, price, all orders, all selling
+    public Mono<Product> getProduct(String itemNumber) {
+        SdkPublisher<Page<DBOrder>> ordersByItemNumber = orderRepository.getOrdersByItemNumber(itemNumber);
+        SdkPublisher<Page<DBReceipt>> soldItemsByItemNumber = sellingRepository.getSoldItemsByItemNumber(itemNumber);
+
+        Flux<DBOrder> ordersFlux = Flux.from(ordersByItemNumber)
+                .log()
+                .map(Page::items)
+                .flatMap(Flux::fromIterable);
+
+        Flux<DBReceipt> receiptsFlux = Flux.from(soldItemsByItemNumber)
+                .log()
+                .map(Page::items)
+                .flatMap(Flux::fromIterable);
+
+        return Mono.zip(ordersFlux.collectList(),
+                        receiptsFlux.collectList()).log()
+                .map(objects -> productMapper.map(itemNumber, objects.getT1(), objects.getT2()));
     }
 
-    public Product getProduct(String itemNumber) {
-        List<DBOrder> orderedProducts = orderService.getOrderedProductsByItemNumber(itemNumber);
-        List<DBReceipt> soldItems = sellingService.getSoldItemsByItemNumber(itemNumber);
-        return productMapper.map(itemNumber, orderedProducts, soldItems);
+    // Update the purchase price, stock information
+    public Mono<Result> updateProduct(ProductStockData productStockData) {
+        DBProduct dbProduct = productMapper.mapProduct(productStockData);
+        return Mono.fromFuture(productRepository.getProductByItemNumber(productStockData.getItemNumber()))
+                .doOnSuccess(Objects::requireNonNull)
+                .doOnNext(product -> productRepository.updateProduct(mergeDbProducts(product, dbProduct)))
+                .doOnSuccess(Objects::requireNonNull)
```

Comment:
`doOn...` methods are for side effects like logging, business logic rarely should be placed there
also `doOnNext` doesn't return the return value of the `updateProduct` method, it just ignores it
consider using `flatMap` here

This is true for all the places where you use `doOnNext`

Reply:
I checked the other places, I used it where data saved into the DB, those methods return with CompletableFuture<Void>, I think in those places I can use doOnNext, then return success/fail